### PR TITLE
fx quant: fix fq when input is quantized and node does not need fq

### PR DESCRIPTION
**Pull request by vkuzo**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #49420 fx quant: make sure observer is inserted before a quantized output
* **#49382 fx quant: fix fq when input is quantized and node does not need fq**
* #49239 fx quant: do not insert observers at quantized inputs
* #49238 fx quant: move {input|output}_quantized_idxs cfg from convert to prepare
* #49213 eager quant: remove fake_quant after add/mul nodes during QAT

Summary:

Fixes an edge case.  If the input to the graph is quantized and the
first node does not need activation observation, makes sure that
the observer is not inserted.

Test Plan:

```
python test/test_quantization.py TestQuantizeFxOps.test_int8_input_no_unnecessary_fq
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25551041](https://our.internmc.facebook.com/intern/diff/D25551041)

### CODE REVIEW

File: torch/quantization/fx/quantize.py [hunk 1]
```diff
@@ -438,6 +438,11 @@ def load_arg(a):
                 cur_placeholder_node_idx = placeholder_node_seen_cnt
                 placeholder_node_seen_cnt += 1
                 if cur_placeholder_node_idx in input_quantized_idxs:
+                    # If an input node is marked as quantized, it was observed
+                    # elsewhere. For now, make it as observed for the purposes
+                    # of this file.  If in the future we will need the observer
+                    # instance for anything, we will have to adjust the API
+                    observed_node_names_set.add(node.name)
```
Comment (jerryzh168) [inline torch/quantization/fx/quantize.py:445]: oh, should this be merged with previous pR

Comment (jerryzh168) [reply torch/quantization/fx/quantize.py:445]: I think we should always do this. What are the cases that needs observer that might come up in the future?

Comment (vkuzo) [reply torch/quantization/fx/quantize.py:441]: > I think we should always do this. What are the cases that needs observer that might come up in the future?

np, removed the comment

> oh, should this be merged with previous pR

I don't feel too strongly, but technically this is a separate bug and fix.  The previous PR removes the observer on the input of the first node, if the input is quantized.  This PR removes the observer on the output of the first node, if the input is quantized and the node does not need observation.  Lmk if you prefer them to be combined.

Comment (jerryzh168) [reply torch/quantization/fx/quantize.py:441]: sg, I don't have a strong preference either.

File: test/quantization/test_quantize_fx.py [hunk 2]
```diff
@@ -1524,6 +1524,32 @@ def forward(self, x):
...
+            def __init__(self, scalar):
+                super().__init__()
+                self.scalar = scalar
+                self.add_func = torch.nn.quantized.FloatFunctional()
+
+            def forward(self, x):
+                return self.add_func.add_scalar(x, self.scalar)
+
+        m = M(0.5)
+        d = {}
```
Comment (jerryzh168) [inline test/quantization/test_quantize_fx.py:1543]: remove?

**Review by jerryzh168 — APPROVED** [review]

**Pull request closed without merging**